### DISCUSSION TITLE
[help] What’s the best way to load data using `data()` with targets?

### DISCUSSION THREAD

**Discussion by b-rodrigues** [ropensci/targets#1038] [Help] [answered]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Hi

because tar_target() needs pure functions, loading data with `data()` can have an unexpected behaviour for new users (the resulting target is the dataset’s name as a string). I use the following function to load data:

```
read_data <- function(data_name, package_name){
  data(list = data_name, package = package_name)
  get(data_name)
}
```

where both `data_name` and `package_name` are characters. I was wondering if there was perhaps a simpler and recommended way of doing this? I’ve looked around on the internet, but can always only find advice on reading in data from disk.

**Comment by wlandau** [comment, marked as answer]
I recommend setting the `envir` argument of `data()`, then extracting the data from that environment.

```r
read_data <- function(data_name, package_name) {
  temp <- new.env(parent = emptyenv())
  data(list = data_name, package = package_name, envir = temp)
  get(data_name, envir = temp)
}
```

**Comment by b-rodrigues** [comment]
Thanks for the answer! Could you explain why it’s preferable to set the `envir` argument?

**Reply by wlandau** [reply]
The `envir` argument of `data()` defaults to the global environment, and that is where the trouble comes from. If you set `envir` to a local temporary environment and get the data object, the function can avoid side effects and just return a value.

**Reply by b-rodrigues** [reply]
Many thanks for your help and clarifications!
